pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
+    filter_backends = (DjangoFilterBackend, SearchFilter, OrderingFilter)
+    filter_fields = ('user__id', 'actor__id', 'active')
+    frozen_fields = ('id', 'actor', 'inserted_at', 'user', 'ended_at')
+    frozen_on_create = ('ended_at', 'end_reason', 'active', 'inserted_at')
 
-    def update(self, request, *args, **kwargs):
+    def create(self, request, *args, **kwargs):
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
 
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id).exists():
+            raise ValidationError({'active': ['There can only be one active nomination.']})
+
+        serializer = self.get_serializer(
+            data=ChainMap(
+                request.data,
+                {"active": True}
+            )
+        )
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def partial_update(self, request, *args, **kwargs):
+        """
+        DRF method for updating a Nomination.
+
+        Called by the Django Rest Framework in response to the corresponding HTTP request.
+        """
         for field in request.data:
             if field in self.frozen_fields:
                 raise ValidationError({field: ['This field cannot be updated.']})
 
         instance = self.get_object()
         serializer = self.get_serializer(instance, data=request.data, partial=True)
         serializer.is_valid(raise_exception=True)
+
+        data = serializer.validated_data
+
+        # There are three distinct PATCH scenarios we need to validate.
+        if instance.active and 'active' not in data:
+            # 1. We're updating an active nomination without ending it.
+            if request.data.get('end_reason'):
+                raise ValidationError(
+                    {'end_reason': ["An active nomination can't have an end reason."]}
+                )
+
+        elif instance.active and not data.get('active'):
+            # 2. We're ending an active nomination.
+            if 'reason' in data:
+                raise ValidationError(
+                    {'reason': ['This field cannot be set when ending a nomination.']}
+                )
+
+            if 'end_reason' not in request.data:
+                raise ValidationError(
+                    {'end_reason': ['This field is required when ending a nomination.']}
+                )
+
+            instance.ended_at = timezone.now()
+
+        elif 'active' in data:
```

Comment:
It should only matter when `active` is `True`. We shouldn't care if a user redundantly sets it to `False` when the nomination is already inactive.

```suggestion
        elif data.get('active'):
```

Reply:
I don't agree with this.

While we've merged the `/end` endpoint into this `PATCH` endpoint, I still think there are two semantically different operations: Updating a reason, which has no functional/semantic consequences, and ending a nomination, which does create a functional/semantic difference. Since the `active` field can only be changed by the latter, I think that it's not right to ignore it implicitly when performing an update reason operation.

If we do allow an implicit ignore of `active=False`, we also have no way of determining the intention of the user when they send a request like :

```python
{
    'active': False,
    'end_reason': 'An end_reason',
}
```
While this may seem harmless, it creates a race condition if two different users try to end the same nomination at roughly the same time. The request that's processed first will end the nomination, setting the initial `end_reason`, but one that's processed second will overwrite the `end_reason` of the former as the endpoint has no way of determining whether or not it's an update or a request that tries to end an already ended nomination. 

Therefore, I think it's better that we invalidate requests that would update a field, but also carry a field that should never be part of an update in the first place.

Reply:
Okay, good point on the race condition. I propose this to make the errors clearer:

```py
active = data.get('active')
if instance.active:
    has_reason = 'end_reason' in data

    if active is False and not has_reason:
        raise ValidationError(
            {'end_reason': ['This field is required when ending a nomination.']}
        )
    elif has_reason:
        raise ValidationError(
            {'end_reason': ["An active nomination can't have an end reason."]}
        )
elif active:
    raise ValidationError(
        {'active': ["Cannot make an inactive nomination active again."]}
    )
elif active is False:
    raise ValidationError(
        {'active': ["The nomination is already inactive."]}
    )
```

Reply:
I like the old version better, since it has a separate "top-level" conditional block for the end_nomination action, which also needs to set the `instance.ended_at` field. 

It could be included here as an additional `active is False and has_reason:` in the nested conditional block in `if instance.active` or, since it's not validation, in a separate `if` statement after the validation `if-statement, but I'm not sure if that would make it easier to follow.